### PR TITLE
Cleaned up image classification cpp example

### PR DESCRIPTION
**Pull request by lebeg**
## Description ##

Cleaned up image classification cpp example.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete

### Changes ###
- Introduced mxnet build option
- Enabled build on mac
- Cleaned up cmake file
- Applied minor code improvements
- Applied code formatting

#### Code improvements ####
- std::string -> const std::string& as method or function parameters
- Handling buffer memory with std::unique_ptr<char[]> instead of raw pointer
- Use std::size_t instead of int for sizes where possible
- Fixed log phrasing and whitespace formatting
- Used ostream for logs instead of printf
- Used standard exit codes from main
- Added license header

### Tested ###
- Mac OS
- Ubuntu
- Windows

### CODE REVIEW

File: example/image-classification/predict-cpp/image-classification-predict.cc [hunk 1]
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
...
-        if (buffer_) {
-          delete[] buffer_;
-          buffer_ = NULL;
-        }
+  char* GetBuffer() {
+    return buffer_;
+  }
+
+  ~BufferFile() {
+    if (buffer_) {
```
Comment (larroy) [inline example/image-classification/predict-cpp/image-classification-predict.cc]: the check is not needed

Comment (larroy) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Better a smart pointer in any case?

File: example/image-classification/predict-cpp/image-classification-predict.cc [hunk 2]
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
...
+      *ptr_image_r++ = static_cast<mx_float>(*data++) - mean_r;;
     }
+  }
 }
 
 // LoadSynsets
 // Code from : https://github.com/pertusa/mxnet_predict_cc/blob/master/mxnet_predict.cc
-std::vector<std::string> LoadSynset(std::string synset_file) {
-    std::ifstream fi(synset_file.c_str());
+std::vector <std::string> LoadSynset(std::string synset_file) {
```
Comment (larroy) [inline example/image-classification/predict-cpp/image-classification-predict.cc]: why space before template arg?

File: example/image-classification/predict-cpp/image-classification-predict.cc [hunk 3]
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
...
-        std::cerr << "Error opening synset file " << synset_file << std::endl;
-        assert(false);
-    }
+  if (!fi.is_open()) {
+    std::cerr << "Error opening synset file " << synset_file << std::endl;
+    assert(false);
+  }
 
-    std::vector<std::string> output;
+  std::vector <std::string> output;
```
Comment (larroy) [inline example/image-classification/predict-cpp/image-classification-predict.cc]: is this your IDE settings?

**Comment by piiswrong** [comment]
Those keys are release signing keys. If you are not going to drive a release you don't need it

File: CMakeLists.txt [hunk 4]
```diff
@@ -671,8 +672,7 @@ if(USE_CPP_PACKAGE)
   add_subdirectory(cpp-package)
 endif()
 
-# Problems on Mac OS X: 1. librt not available 2. mxnet built as MODULE library, which can't be linked.
```
Comment (marcoabreu) [inline CMakeLists.txt]: Has this been resolved?

File: KEYS [hunk 5]
```diff
@@ -480,3 +480,40 @@ IxR3jVTKye+UerEtN8yATW8CRIKO3IobUfLMDdPCLO7uzoW95cI35Y0l8JgK2NeU
 =E4W5
 -----END PGP PUBLIC KEY BLOCK-----
 
+pub   rsa2048 2018-02-07 [SC] [expires: 2020-02-07]
```
Comment (marcoabreu) [inline KEYS]: Since you're not going to make a release, I don't think that your key should be in here.

**Comment by marcoabreu** [comment]
Could you please add on which platforms you have tested your modifications? Since we're not testing examples on CI, we'll have to test them ourselves.

File: CMakeLists.txt (hunk 4 above)
Comment (lebeg) [reply CMakeLists.txt]: Yes

**Comment by lebeg** [comment]
Listed tested platforms in the description

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 3 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Thanks, this was a weird default setting on a new machine

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 1 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Done

**Comment by marcoabreu** [comment]
Could you please also test Windows?

**Comment by lebeg** [comment]
Added windows to tested platforms

File: example/image-classification/predict-cpp/CMakeLists.txt [hunk 6]
```diff
@@ -1,34 +1,28 @@
...
-    find_package(OpenCV REQUIRED COMPONENTS core highgui imgproc)
-  endif()
-
-  if(NOT MSVC)
-    set(UNITTEST_STATIC_LINK ON)
-  endif()
+# ---[ OpenCV
+if(NOT USE_OPENCV OR NOT OpenCV_FOUND)
+  message(WARNING "\
+OpenCV should be enabled and found to build image classification example, skipping...")
```
Comment (marcoabreu) [inline example/image-classification/predict-cpp/CMakeLists.txt]: This is a critical error, right?

Comment (lebeg) [reply example/image-classification/predict-cpp/CMakeLists.txt]: The example will just not be build and it will not break the whole project configuration.

**Comment by lebeg** [comment]
@marcoabreu @piiswrong @cjolivier01 @szha do you want to take another look and merge if possible?

**Comment by marcoabreu** [comment]
I can't really help here, sorry.

File: CMakeLists.txt [hunk 7]
```diff
@@ -32,6 +32,7 @@ mxnet_option(USE_GPROF            "Compile with gprof (profiling) flag" OFF)
 mxnet_option(USE_CXX14_IF_AVAILABLE "Build with C++14 if the compiler supports it" OFF)
 mxnet_option(USE_VTUNE            "Enable use of Intel Amplifier XE (VTune)" OFF) # one could set VTUNE_ROOT for search path
 mxnet_option(ENABLE_CUDA_RTC      "Build with CUDA runtime compilation support" ON)
+mxnet_option(BUILD_CPP_EXAMPLES   "Build cpp examples" OFF)
```
Comment (cjolivier01) [inline CMakeLists.txt]: default to on, please. Otherwise, people may check in/merge broken code.

File: example/image-classification/predict-cpp/CMakeLists.txt (hunk 6 above)
Comment (cjolivier01) [reply example/image-classification/predict-cpp/CMakeLists.txt]: This certainly is chopping out a lot of stuff.  WHat's the net change in static linking logic?

File: example/image-classification/predict-cpp/image-classification-predict.cc [hunk 8]
```diff
@@ -1,283 +1,298 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
 /*!
- *  Copyright (c) 2015 by Xiao Liu, pertusa, caprice-j
```
Comment (cjolivier01) [inline example/image-classification/predict-cpp/image-classification-predict.cc]: I don't think you can do this, can you? I was under the impression that license text could never be removed.
  @smarthi , this this ok?

File: example/image-classification/predict-cpp/image-classification-predict.cc [hunk 9]
```diff
@@ -1,283 +1,298 @@
...
-
+#include <memory>
+#include <iomanip>
 #include <opencv2/opencv.hpp>
+// Path for c_predict_api
+#include <mxnet/c_predict_api.h>
 
 const mx_float DEFAULT_MEAN = 117.0;
 
+std::string trim(const std::string& input) {
```
Comment (cjolivier01) [inline example/image-classification/predict-cpp/image-classification-predict.cc]: make static, please

**Comment by cjolivier01** [comment]
Can you please summarize the "cleaning up" items in the description? It's a little hard to follow what the added value is from the diff, since it's a bit scattered.

File: CMakeLists.txt (hunk 7 above)
Comment (lebeg) [reply CMakeLists.txt]: Ok

File: example/image-classification/predict-cpp/CMakeLists.txt (hunk 6 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/CMakeLists.txt]: It is not about static, it's an early return from this cmake file. The example can not be built if there is no OpenCV and will not be added to the main project. That's it.

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 8 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Technically I'm not removing contributors, just naming them as a group. I followed the pattern in the rest of the project, but will be happy to change it back.

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 9 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Ok

**Comment by lebeg** [comment]
@cjolivier01 The changes are already listed:

### Changes ###
* Introduced mxnet build option
* Enabled build on mac
* Cleaned up cmake file
* Applied minor code improvements
* Applied code formatting

**Comment by lebeg** [comment]
@cjolivier01 please take another look, thanks!

File: example/image-classification/predict-cpp/CMakeLists.txt (hunk 6 above)
Comment (cjolivier01) [reply example/image-classification/predict-cpp/CMakeLists.txt]: isn’t the unittest static link flag removed?

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 8 above)
Comment (cjolivier01) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: is the guy’s name still there?

**Comment by cjolivier01** [comment]
Can you please summarize the "cleaning up" items in the description? It's a little hard to follow what the added value is from the diff, since it's a bit scattered.

File: example/image-classification/predict-cpp/image-classification-predict.cc (hunk 8 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/image-classification-predict.cc]: Yes, brought it back

File: example/image-classification/predict-cpp/CMakeLists.txt (hunk 6 above)
Comment (cjolivier01) [reply example/image-classification/predict-cpp/CMakeLists.txt]: UNITTEST_STATIC_LINK is descriptive and editable.  Changing the check to MSVC at the actual interpretation point makes it difficult to see what's being done.
Also, I think MSVC has to build static because of dllimport issues

**Comment by cjolivier01** [comment]
Also, is this used in any of the unit tests or nightly tests?  If not, probably we should be running at least one test with it.

**Comment by lebeg** [comment]
@cjolivier01 for now this is not tested on ci, but I will ingerate this as the base for valgrind memcheck tests

File: example/image-classification/predict-cpp/CMakeLists.txt (hunk 6 above)
Comment (lebeg) [reply example/image-classification/predict-cpp/CMakeLists.txt]: I can leave the variable itself but would rename it, because it's an example, not a unit test.

**Comment by lebeg** [comment]
Added 'Code improvements' section for detailisation
